download_ta_bdys.py

```diff
@@ -271,14 +255,192 @@ def main():
+        sys.exit(1)
+
+    try:
+       count_json = json.load(count_response)
+       feat_count = int(count_json['count'])
+    except Exception, e:
+       logger.warning('Could not get feature count for uri %s: %s' % (count_uri, str(e)))
+
+    logger.info('Your OGR version is: ' + str(version_num))
+    logger.info('Stats NZ ArcGIS server version is : ' + str(server_version))
+    logger.info('Stats NZ ArcGIS server pagination flag is : ' + str(server_pagination))
+
+    # Temp layer to store features
+    tmp_drv = ogr.GetDriverByName("MEMORY")
+    tmp_ds = tmp_drv.CreateDataSource('memDs')
+    tmp_opn = tmp_drv.Open('memDs',1)
+    input_lyr = tmp_ds.CreateLayer('tmpLyr',srs=layer_output_srs,geom_type=ogr.wkbMultiPolygon)
+
+    # Use OGR to do paging if version 2.x and server supports paging 
+    if version_num >= 2000000 \
+        and server_version >= 10.3  \
+        and str(server_pagination).lower() == 'true' \
+        and feat_count \
+        and display_field:
+
+        for i in range(feat_count):
+
+            offset = round_up(i, page_size)
+
+            page_uri = base_uri + '/' + str(latest_layer) + \
+                 '/query?f=json&where=1=1&returnGeometry=true&outSR=' + str(srid) + \
+                 'orderByFields=' + display_field + '&resultOffset=' + offset + '&resultRecordCount=' + feat_count 
+
+            try:
+                geojs = geojson_drv.Open(page_uri)
+            except Exception, e:
+                logger.fatal('Could not read geojson from uri %s: %s' % (page_uri, str(e)))
+                sys.exit(1)
+
+            try:
+                geojs_lyr = geojs.GetLayer(0)
+            except Exception, e:
+                logger.fatal('Could not get layer from geojson: %s' % (str(e)))
+                sys.exit(1)
+
+            for feature in geojs_lyr:
+                if not tmp_feat_defn:
+                    tmp_feat_defn = feature.GetDefnRef()
+                    for i in range(tmp_feat_defn.GetFieldCount()):
+                       field_defn = tmp_feat_defn.GetFieldDefn(i)
+                       input_lyr.CreateField(field_defn)
+
+                input_lyr.CreateFeature(feature)
+
+    # Otherwise use filtering to do paging
+    else:
+
+        # Get a list of Object IDs
+        object_ids_uri = base_uri + '/' + str(latest_layer) + \
+                '/query?f=json&where=1=1&returnGeometry=False&outFields=objectid'
+
+        try:
+            response = urllib2.urlopen(object_ids_uri)
+        except Exception, e:
+            logger.fatal('Could open uri %s: %s' % (object_ids_uri, str(e)))
+            sys.exit(1)
+
+        try:
+            json_ids = json.load(response)
+        except Exception, e:
+            logger.fatal('Could load json from uri %s: %s' % (object_ids_uri, str(e)))
+            sys.exit(1)
+
+        # Build page requests by getting lists of ObjectIDs 
+        pages = []
+        for i in range(len(json_ids['features'])):
```

Comment:
Minor tip: I would say you don't need to loop twice for this paging code (i.e once to determine page list subsets and then once to loop over the subsets themselves. You should be able to use python list slicing or something similar to reduce this to one loop.